internal/jujuclient/dial.go

```
@@ -348,6 +348,11 @@ func (c *Connection) ConnectStream(path string, attrs url.Values) (base.Stream,
 	if err != nil {
 		return nil, errors.E(op, err)
 	}
+	ok = names.IsValidUser(user)
+	if !ok {
+		zapctx.Error(c.ctx, "invalid controller credentials", zap.String("user", user))
+		return nil, errors.E(op, "invalid/missing controller credentials")
+	}
```

This is an unrelated change I had sitting locally. You can technically hit a panic without this change if the controller's credentials are not present in Vault due to the `names.NewUserTag()` function. So I've just added some validation.